beacon_node/beacon_chain/src/builder.rs

```diff
@@ -412,43 +423,49 @@ where
 
-        // Check that the given state lies on an epoch boundary. Due to the database only storing
-        // full states on epoch boundaries and at restore points it would be difficult to support
-        // starting from a mid-epoch state.
-        if weak_subj_slot % TEthSpec::slots_per_epoch() != 0 {
-            return Err(format!(
-                "Checkpoint state at slot {} is not aligned to epoch start. \
-                 Please supply an aligned checkpoint with state.slot % 32 == 0",
-                weak_subj_slot,
-            ));
+        // Ensure the state is advanced to an epoch boundary.
+        let slots_per_epoch = TEthSpec::slots_per_epoch();
+        if weak_subj_state.slot() % slots_per_epoch != 0 {
+            debug!(
+                log,
+                "Advancing checkpoint state to boundary";
+                "state_slot" => weak_subj_state.slot(),
+                "block_slot" => weak_subj_block.slot(),
+            );
+            while weak_subj_state.slot() % slots_per_epoch != 0 {
+                per_slot_processing(&mut weak_subj_state, None, &self.spec)
+                    .map_err(|e| format!("Error advancing state: {e:?}"))?;
+            }
         }
 
         // Prime all caches before storing the state in the database and computing the tree hash
         // root.
         weak_subj_state
             .build_caches(&self.spec)
             .map_err(|e| format!("Error building caches on checkpoint state: {e:?}"))?;
-        weak_subj_state
+        let weak_subj_state_root = weak_subj_state
             .update_tree_hash_cache()
             .map_err(|e| format!("Error computing checkpoint state root: {:?}", e))?;
 
-        let latest_block_slot = weak_subj_state.latest_block_header().slot;
+        let weak_subj_slot = weak_subj_state.slot();
+        let weak_subj_block_root = weak_subj_block.canonical_root();
 
-        // We can only validate the block root if it exists in the state. We can't calculated it
-        // from the `latest_block_header` because the state root might be set to the zero hash.
-        if let Ok(state_slot_block_root) = weak_subj_state.get_block_root(latest_block_slot) {
-            if weak_subj_block_root != *state_slot_block_root {
-                return Err(format!(
-                    "Snapshot state's most recent block root does not match block, expected: {:?}, got: {:?}",
-                    weak_subj_block_root, state_slot_block_root
-                ));
-            }
+        // Validate the state's `latest_block_header` against the checkpoint block.
+        let state_latest_block_root = weak_subj_state.get_latest_block_root(weak_subj_state_root);
+        if weak_subj_block_root != state_latest_block_root {
+            return Err(format!(
+                "Snapshot state's most recent block root does not match block, expected: {:?}, got: {:?}",
+                weak_subj_block_root, state_latest_block_root
+            ));
```

Comment:
This was the conflicting bit, but should be fine to accept `unstable` changes, as it was built on top of the same checkpoint sync alignment [branch](https://github.com/sigp/lighthouse/pull/4610/commits/2c8f07ed5a853951ec8b26b2f2cf4ef9d210f08c).

Reply:
Agreed. Looks like you faithfully maintained the logic from `unstable` (based on the diff here: https://github.com/sigp/lighthouse/compare/unstable...jimmygchen:lighthouse:merge-unstable-to-deneb-20230822#diff-44c324564286d739ad29b047b484dc7fd1ba47741a3e07665302dcd10ab83f9c)